[expo/dom] Add `source` to `DOMProps` to enable pre-bundled DOM Components

krystofwoldrich opened the pull request:
# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

Changing the DOM Component `source` enables pre building DOM Components.

- used in https://github.com/expo/expo/pull/39958

I've added the type to `DOMProps` ~as it should not be exposed to the users (at least not now)~.

~I've also consider `sourceOverride` prop to avoid users accidentally passing `source`. But all other props use the webview props naming, so this PR sticks to it.~

# Test Plan

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

```jsx
export default function App() {
  return (
    <DOMComponent
      dom={{
        // The provided uri will be rendered instead of the `Hello, world!`
        overrideUri: 'https://docs.expo.dev'
      }}
    />
  );
}
```

```jsx
'use dom';

export default function DOMComponent({}: { dom: import('expo/dom').DOMProps }) {
  return (
    <div>
      <h1>Hello, world!</h1>
    </div>
  );
}
```

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

Kudo reviewed (changes requested):
i actually like `sourceOverride` or even `overrideUri` that acts as `source: { uri: overrideUri }`.

my concern is to prevent people from shooting themselves in the foot. the `source` is just to easy to abuse

krystofwoldrich commented:
Thanks @Kudo, I had the same concern, I've updated the PR with `overrideUri`.

Let me know if like this it's good, or would you also remove it from the public type (and move to internal)?

Kudo commented on packages/expo/src/dom/webview-wrapper.tsx:
```diff
@@ -3,7 +3,7 @@ import React from 'react';
 
 import { getBaseURL } from './base';
-import type { BridgeMessage, DOMProps, WebViewProps, WebViewRef } from './dom.types';
+import type { BridgeMessage, WebViewProps, WebViewRef, DOMProps } from './dom.types';
```
```suggestion
import type { BridgeMessage, DOMProps, WebViewProps, WebViewRef } from './dom.types';
```

not sure what's the intention to put DOMProps at last

krystofwoldrich replied:
No intention, only left over from adding DOMPropsInternal in previous revision.

Kudo commented on packages/expo/src/dom/webview-wrapper.tsx:
```diff
@@ -66,7 +67,7 @@ const RawWebView = React.forwardRef<object, Props>((props, ref) => {
   const webviewRef = React.useRef<WebViewRef>(null);
   const domImperativeHandlePropsRef = React.useRef<string[]>([]);
-  const source = { uri: `${getBaseURL()}/${filePath}` };
+  const source = { uri: overrideUri ??`${getBaseURL()}/${filePath}` };
```
```suggestion
  const source = { uri: overrideUri ?? `${getBaseURL()}/${filePath}` };
```

Kudo reviewed (approved):
may also need to make sure check-packages ci is passed before merge

krystofwoldrich committed:
[internal] Add `source` to `DOMPropsInternal` to enable pre-bundled DOM Components

Update packages/expo/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

move source to DOMProps

use overrideUri

fix lint
